File: TranslationFiestaFlutter/lib/data/repositories/translation_repository_impl.dart
library;

import 'package:http/http.dart' as http;
import '../../core/errors/either.dart';
import '../../core/errors/failure.dart';
import '../../core/utils/logger.dart';
import '../../domain/entities/translation.dart';
import '../../domain/repositories/translation_repository.dart';
import '../services/retry_service.dart';
import '../services/translation_service.dart';

/// Implementation of TranslationRepository interface
/// Single Responsibility: Coordinate translation operations with services
class TranslationRepositoryImpl implements TranslationRepository {
  final UnofficialGoogleTranslateService _unofficialService;
  final RetryService _retryService;
  final Logger logger = Logger.instance;

  TranslationRepositoryImpl(http.Client httpClient)
      : _unofficialService = UnofficialGoogleTranslateService(httpClient),
        _retryService = RetryService();

  @override
  String get serviceName => 'Translation Repository';

  @override
  bool get isConfigured => true;

  @override
  Future<Result<TranslationResult>> translateText(
    TranslationRequest request,
    ApiConfiguration config,
  ) async {
    return _retryService.executeWithRetry(
      () => _translateWithService(request, config),
      config,
      operationName: 'Translation',
      statusCallback: (message) {}, // No callback needed for single translation
    );
  }

  /// Internal method to run translation service
  Future<Result<TranslationResult>> _translateWithService(
    TranslationRequest request,
    ApiConfiguration config,
  ) async {
    logger.info('Starting translation with config: $config');
    logger.info('Using service: ${_unofficialService.serviceName}');

    final result = await _unofficialService.translate(request, config);

    logger.info('Service result: ${result.isRight ? "Success" : "Failure"}');
    if (result.isLeft) {
      logger.error('Service failure: ${result.left.message}');
    }

    return result;
  }

  @override
  Future<Result<BackTranslationResult>> performBackTranslation(
    String text,
    ApiConfiguration config, {
    String intermediateLanguage = 'ja',
  }) async {
    logger.info('Repository: performBackTranslation called');
    logger.info('Repository: text: "$text"');
    logger.info('Repository: config: $config');
    logger.info('Repository: intermediateLanguage: $intermediateLanguage');

    if (text.trim().isEmpty) {
      logger.info('Repository: text is empty, returning empty result');
      return Right(
        BackTranslationResult(
          originalText: text,
          intermediateTranslation: TranslationResult(
            originalText: text,
            translatedText: '',
            sourceLanguage: 'en',
            targetLanguage: intermediateLanguage,
          ),
          finalTranslation: TranslationResult(
            originalText: '',
            translatedText: '',
            sourceLanguage: intermediateLanguage,
            targetLanguage: 'en',
          ),
          timestamp: DateTime.now(),
          totalDuration: Duration.zero,
        ),
      );
    }

    // First translation: source -> intermediate
    final firstRequest = TranslationRequest(
      text: text,
      sourceLanguage: 'en',
      targetLanguage: intermediateLanguage,
    );

    Future<Result<TranslationResult>> firstTranslation() =>
        _translateWithService(firstRequest, config);

    // Second translation: intermediate -> source
    Future<Result<TranslationResult>> secondTranslation(
      String intermediateText,
    ) async {
      final secondRequest = TranslationRequest(
        text: intermediateText,
        sourceLanguage: intermediateLanguage,
        targetLanguage: 'en',
      );
      return _translateWithService(secondRequest, config);
    }

    logger.info('Repository: calling retry service');
    final result = await _retryService.executeBackTranslationWithRetry(
      firstTranslation,
      secondTranslation,
      text,
      config,
      intermediateLanguage: intermediateLanguage,
      statusCallback: (message) {
        logger.info('Repository: status callback: $message');
      },
    );
    logger.info(
      'Repository: retry service result: ${result.isRight ? "Success" : "Failure"}',
    );
    return result;
  }

  @override
  Future<Result<String>> detectLanguage(
    String text,
    ApiConfiguration config,
  ) async {
    logger.debug('Detect language provider: ${config.providerId}');
    if (text.trim().isEmpty) {
      return Left(AppFailure(message: 'Text is empty'));
    }
    return const Right('en');
  }
}

/// Factory for creating TranslationRepository instances
class TranslationRepositoryFactory {
  static TranslationRepository create() {
    final httpClient = http.Client();
    return TranslationRepositoryImpl(httpClient);
  }

  /// Create with custom HTTP client (useful for testing)
  static TranslationRepository createWithClient(http.Client client) {
    return TranslationRepositoryImpl(client);
  }
}

File: TranslationFiestaGo/internal/data/repositories/translation_repository_impl.go
package repositories

import (
	"context"
	"time"
	"translationfiestago/internal/data/services"
	"translationfiestago/internal/domain/entities"
	"translationfiestago/internal/domain/repositories"
	"translationfiestago/internal/utils"
)

// TranslationRepositoryImpl implements the TranslationRepository interface
type TranslationRepositoryImpl struct {
	translationService *services.TranslationService
	logger             *utils.Logger
}

// NewTranslationRepository creates a new translation repository
func NewTranslationRepository(translationMemoryPath string) repositories.TranslationRepository {
	return &TranslationRepositoryImpl{
		translationService: services.NewTranslationService(translationMemoryPath),
		logger:             utils.GetLogger(),
	}
}

// Translate performs a single translation
func (r *TranslationRepositoryImpl) Translate(ctx context.Context, request entities.TranslationRequest) (*entities.TranslationResult, error) {
	r.logger.Debug("Translating text: %s -> %s", request.SourceLang, request.TargetLang)

	var result *entities.TranslationResult
	var err error

	result, err = r.translationService.TranslateUnofficial(ctx, request.Text, request.SourceLang, request.TargetLang)

	if err != nil {
		r.logger.Error("Translation failed: %v", err)
		return nil, err
	}

	// Fill in the missing fields
	result.OriginalText = request.Text
	result.SourceLang = request.SourceLang
	result.TargetLang = request.TargetLang

	return result, nil
}

// TranslateUnofficial performs translation using the unofficial Google Translate API
func (r *TranslationRepositoryImpl) TranslateUnofficial(ctx context.Context, text, sourceLang, targetLang string) (*entities.TranslationResult, error) {
	result, err := r.translationService.TranslateUnofficial(ctx, text, sourceLang, targetLang)
	if err != nil {
		return nil, err
	}

	result.OriginalText = text
	result.SourceLang = sourceLang
	result.TargetLang = targetLang

	return result, nil
}

// BackTranslate performs a full back-translation (source -> intermediate -> source)
func (r *TranslationRepositoryImpl) BackTranslate(ctx context.Context, text, sourceLang, intermediateLang, providerID string) (*entities.BackTranslation, error) {
	r.logger.Info("Starting back-translation: %s -> %s -> %s", sourceLang, intermediateLang, sourceLang)

	startTime := time.Now()

	backTranslation := &entities.BackTranslation{
		Input:            text,
		SourceLang:       sourceLang,
		IntermediateLang: intermediateLang,
		FinalLang:        sourceLang,
		Timestamp:        startTime,
	}

	// Step 1: Translate source -> intermediate
	r.logger.Debug("Step 1: %s -> %s", sourceLang, intermediateLang)

	step1Result, err := r.Translate(ctx, entities.TranslationRequest{
		Text:       text,
		SourceLang: sourceLang,
		TargetLang: intermediateLang,
		ProviderID: providerID,
	})

	if err != nil {
		backTranslation.Error = err
		r.logger.Error("Back-translation step 1 failed: %v", err)
		return backTranslation, err
	}

	backTranslation.Intermediate = step1Result.TranslatedText
	r.logger.Info("Step 1 completed: %d chars", len(backTranslation.Intermediate))

	// Step 2: Translate intermediate -> source (back to original language)
	r.logger.Debug("Step 2: %s -> %s", intermediateLang, sourceLang)

	step2Result, err := r.Translate(ctx, entities.TranslationRequest{
		Text:       backTranslation.Intermediate,
		SourceLang: intermediateLang,
		TargetLang: sourceLang,
		ProviderID: providerID,
	})

	if err != nil {
		backTranslation.Error = err
		r.logger.Error("Back-translation step 2 failed: %v", err)
		return backTranslation, err
	}

	backTranslation.Result = step2Result.TranslatedText
	backTranslation.Duration = time.Since(startTime)

	r.logger.Info("Back-translation completed successfully: %d -> %d -> %d chars (took %v)",
		len(text), len(backTranslation.Intermediate), len(backTranslation.Result), backTranslation.Duration)

	return backTranslation, nil
}
